File: src/algorithms/sorting/selection/double-selection-sort/sources/double_selection_sort_test.py
import importlib

double_selection_sort_module = importlib.import_module("double-selection-sort")
double_selection_sort = double_selection_sort_module.double_selection_sort


def test_sorts_unsorted_array():
    assert double_selection_sort([64, 34, 25, 12, 22, 11, 90]) == [11, 12, 22, 25, 34, 64, 90]


def test_handles_already_sorted_array():
    assert double_selection_sort([1, 2, 3, 4, 5]) == [1, 2, 3, 4, 5]


def test_handles_reverse_sorted_array():
    assert double_selection_sort([5, 4, 3, 2, 1]) == [1, 2, 3, 4, 5]


def test_handles_array_with_duplicate_values():
    assert double_selection_sort([3, 1, 4, 1, 5, 9, 2, 6, 5]) == [1, 1, 2, 3, 4, 5, 5, 6, 9]


def test_handles_single_element_array():
    assert double_selection_sort([42]) == [42]


def test_handles_empty_array():
    assert double_selection_sort([]) == []


def test_handles_array_with_negative_numbers():
    assert double_selection_sort([3, -1, 0, -5, 2]) == [-5, -1, 0, 2, 3]


def test_handles_even_length_array():
    assert double_selection_sort([4, 2, 6, 1]) == [1, 2, 4, 6]


def test_does_not_mutate_original_array():
    original = [3, 1, 2]
    sorted_result = double_selection_sort(original)
    assert sorted_result == [1, 2, 3]
    assert original == [3, 1, 2]


if __name__ == "__main__":
    test_sorts_unsorted_array()
    test_handles_already_sorted_array()
    test_handles_reverse_sorted_array()
    test_handles_array_with_duplicate_values()
    test_handles_single_element_array()
    test_handles_empty_array()
    test_handles_array_with_negative_numbers()
    test_handles_even_length_array()
    test_does_not_mutate_original_array()
    print("All tests passed!")

File: src/algorithms/sorting/selection/double-selection-sort/double_selection_sort_test.go
package main

import (
	"reflect"
	"testing"
)

func TestSortsUnsortedArray(t *testing.T) {
	result := doubleSelectionSort([]int{64, 34, 25, 12, 22, 11, 90})
	expected := []int{11, 12, 22, 25, 34, 64, 90}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("expected %v, got %v", expected, result)
	}
}

func TestHandlesAlreadySortedArray(t *testing.T) {
	result := doubleSelectionSort([]int{1, 2, 3, 4, 5})
	expected := []int{1, 2, 3, 4, 5}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("expected %v, got %v", expected, result)
	}
}

func TestHandlesReverseSortedArray(t *testing.T) {
	result := doubleSelectionSort([]int{5, 4, 3, 2, 1})
	expected := []int{1, 2, 3, 4, 5}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("expected %v, got %v", expected, result)
	}
}

func TestHandlesArrayWithDuplicateValues(t *testing.T) {
	result := doubleSelectionSort([]int{3, 1, 4, 1, 5, 9, 2, 6, 5})
	expected := []int{1, 1, 2, 3, 4, 5, 5, 6, 9}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("expected %v, got %v", expected, result)
	}
}

func TestHandlesSingleElementArray(t *testing.T) {
	result := doubleSelectionSort([]int{42})
	expected := []int{42}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("expected %v, got %v", expected, result)
	}
}

func TestHandlesEmptyArray(t *testing.T) {
	result := doubleSelectionSort([]int{})
	if len(result) != 0 {
		t.Errorf("expected empty slice, got %v", result)
	}
}

func TestHandlesArrayWithNegativeNumbers(t *testing.T) {
	result := doubleSelectionSort([]int{3, -1, 0, -5, 2})
	expected := []int{-5, -1, 0, 2, 3}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("expected %v, got %v", expected, result)
	}
}

func TestHandlesEvenLengthArray(t *testing.T) {
	result := doubleSelectionSort([]int{4, 2, 6, 1})
	expected := []int{1, 2, 4, 6}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("expected %v, got %v", expected, result)
	}
}

func TestDoesNotMutateOriginalArray(t *testing.T) {
	original := []int{3, 1, 2}
	originalCopy := []int{3, 1, 2}
	sorted := doubleSelectionSort(original)
	if !reflect.DeepEqual(sorted, []int{1, 2, 3}) {
		t.Errorf("expected sorted [1 2 3], got %v", sorted)
	}
	if !reflect.DeepEqual(original, originalCopy) {
		t.Errorf("original array was mutated: got %v", original)
	}
}
